### PR TITLE
Basic functionality of Delta Table Features

### PR DESCRIPTION
**Pull request by xupefei**
This PR implements Table Features proposed in the feature request [[Feature Request] Table Features to specify the features needed to read/write to a table](https://github.com/delta-io/delta/issues/1408) and the PROTOCOL doc: [Protocol update for Table Features](https://github.com/delta-io/delta/pull/1450).

This PR implements the basic functionality, including
- The protocol structure and necessary APIs
- Protocol upgrade logic
- Append-only feature ported to Table Features
- Protocol upgrade path
- User-facing APIs, such as allowing referencing features manually
- Partial test coverage

Not covered by this PR:
- Adapt all features
- Full test coverage
- Make `DESCRIBE TABLE` show referenced features
- Enable table clone and time travel paths

### Introduction

Table Features support starts from reader protocol version `3` and writer version `7`. When supported, features can be **referenced** by a protocol by placing a `DeltaFeatureDescriptor` into the protocol's `readerFeatures` and/or `writerFeatures`.

A feature can be one of two types: writer-only and reader-writer. The first type means that only writers must care about such a feature, while the latter means that in addition to writers, readers must also be aware of the feature to read the data correctly. We now have the following features released:

- `appendOnly`: legacy, writer-only
- `invariants`: legacy, writer-only
- `checkConstriants`: legacy, writer-only
- `changeDataFeed`: legacy, writer-only
- `generatedColumns`: legacy, writer-only
- `columnMapping`: legacy, reader-writer
- `identityColumn`: legacy, writer-only
- `deletionVector`: native, reader-writer

Some examples of the `protocol` action:

```scala
// Valid protocol. Both reader and writer versions are capable.
Protocol(
  minReaderVersion = 3,
  minWriterVersion = 7,
  readerFeatures = {(columnMapping,enabled), (changeDataFeed,enabled)},
  writerFeatures = {(appendOnly,enabled), (columnMapping,enabled), (changeDataFeed,enabled)})

// Valid protocol. Only writer version is capable. "columnMapping" is implicitly enabled by readers.
Protocol(
  minReaderVersion = 2,
  minWriterVersion = 7,
  readerFeatures = None,
  writerFeatures = {(columnMapping,enabled)})

// Invalid protocol. Reader version does enable "columnMapping" implicitly.
Protocol(
  minReaderVersion = 1,
  minWriterVersion = 7,
  readerFeatures = None,
  writerFeatures = {(columnMapping,enabled)})
```

When reading or writing a table, clients MUST respect all enabled features.

Upon table creation, the system assigns the table a minimum protocol that satisfies all features that are **automatically enabled** in the table's metadata. This means the table can be on a "legacy" protocol with both `readerFeatures` and `writerFeatures` set to `None` (if all active features are legacy, which is the current behavior) or be on a Table Features-capable protocol with all active features explicitly referenced in `readerFeatures` and/or `writerFeatures` (if one of the active features is Table Features-native or the user has specified a Table Features-capable protocol version).

It's possible to upgrade an existing table to support table features. The update can be either for writers or for both readers and writers. During the upgrade, the system will explicitly reference all legacy features that are implicitly supported by the old protocol.

Users can mark a feature to be required by a table by using the following commands:
```sql
-- for an existing table
ALTER TABLE table_name SET TBLPROPERTIES (delta.feature.featureName = 'enabled')
-- for a new table
CREATE TABLE table_name ... TBLPROPERTIES (delta.feature.featureName = 'enabled')
-- for all new tables
SET spark.databricks.delta.properties.defaults.feature.featureName = 'enabled'
```
When some features are set to `enabled` in table properties and some others in Spark sessions, the final table will enable all features defined in two places:
```sql
SET spark.databricks.delta.properties.defaults.feature.featureA = 'enabled';
CREATE TABLE table_name ... TBLPROPERTIES (delta.feature.featureB = 'enabled')
--- 'table_name' will have 'featureA' and 'featureB' enabled.
```

## How was this patch tested?

New unit tests.

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala [hunk 1]
```diff
@@ -286,16 +286,22 @@ class Snapshot(
   lazy val numIndexedCols: Int = DeltaConfigs.DATA_SKIPPING_NUM_INDEXED_COLS.fromMetaData(metadata)
 
   /** Return the set of properties of the table. */
-  def getProperties: mutable.HashMap[String, String] = {
-    val base = new mutable.HashMap[String, String]()
+  def getProperties: mutable.Map[String, String] = {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala:289]: why this change?

File: core/src/main/scala/org/apache/spark/sql/delta/util/Utils.scala [hunk 2]
```diff
@@ -35,4 +35,11 @@ object Utils {
   def getRandomPrefix(numChars: Int): String = {
     Random.alphanumeric.take(numChars).mkString
   }
+
+  /**
+   * Indicates whether Delta is currently running unit tests.
+   */
+  def isTesting: Boolean = {
+    System.getenv("DELTA_TESTING") != null
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/util/Utils.scala:43]: is this set env set any where?

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaLog.scala [hunk 3]
```diff
@@ -329,39 +330,78 @@ class DeltaLog private(
...
+    }
+
+    val clientSupportedFeatureNames = getEnabledFeatures(clientSupportedProtocol).map(_.name)
+    val tableEnabledFeatureNames = getEnabledFeatures(tableProtocol).map(_.name)
+    if (tableEnabledFeatureNames.subsetOf(clientSupportedFeatureNames) &&
+      clientSupportedVersion >= tableRequiredVersion) {
+      return // protocol validation complete
+    }
+
+    val clientUnsupportedFeatureNames =
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaLog.scala:363]: Can you have a few more inline comments explaining this monolithic piece of code.

File: core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+import org.apache.spark.sql.SparkSession
+
+object TableFeatureStore {
+
+  /**
+   * All table features recognized by this client. Update this set when you added a new Table
+   * Feature.
+   */
+  val allSupportedFeaturesMap: Map[String, TableFeature] = {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala:32]: nit: this file is named TableFeatureStore but the only class named that has very little functionality. Instead the main class in this file that everyone will search for is TableFeature. So my suggestion to structure this would be to make the file TableFeature.scala and put this method in TableFeature companion object. Then we dont need any new name like TableFeatureStore file or class.

File: core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        TestLegacyReaderWriterFeature,
+        TestReaderWriterFeature)
+    }
+    val featureMap = features.map(f => f.name.toLowerCase(Locale.ROOT) -> f).toMap
+    require(features.size == featureMap.size, "Lowercase feature names must not duplicate.")
+    featureMap
+  }
+}
+
+/* ---------------------------------------- *
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala:47]: It is actually more common practice to put the base trait/class first and then implementations. See actions.scala or DeltaOperations.scala.

**Review by tdas — CHANGES_REQUESTED** [review]
Overall looks great. Thanks for making such a structured code. I generally dont like PRs that are so big, but they tend to be not structured well and requiring lots of review cycles. But kudos to you for structuring this large PR well. And I can also see that this functionality is hard to break up further. LGTM if you fix the few comments I have.

File: core/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala (hunk 1 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala:289]: Because we want to keep enabled features at the end of the prop list, otherwise too many enabled features will make the output unreadable. To do this we have to use `LinkedHashMap` which does keep the insertion order: https://www.scala-lang.org/api/2.13.6/scala/collection/mutable/LinkedHashMap.html.
Meanwhile we do not want to break the ABI compatibility, thus keeping the return `mutable.HashMap` interface unchanged.

File: core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala (hunk 4 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala]: Thanks for the suggestion! I modified the code following your comment.

File: core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala (hunk 5 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/TableFeatureStore.scala]: Done! Now it's base class/trait, then the companion object, then features.
